### PR TITLE
Rename TARPL to the Rustonomicon

### LINKED ISSUE #27412: Bikeshed "The Advanced Rust Programming Language" as a name
@dherman suggested this iirc -- but I can't remember.

@brson is concerned that it is too similar

Originally it was called *The Unsafe Rust Programming Language* since it largely focuses on unsafe code and the stuff that matters to it.

### PR DESCRIPTION
**Pull request by Gankra**
Closes #27412 

r? @brson

### CODE REVIEW

**Comment by Gankro** [comment]
This is also the final resolution of #13871

**Comment by huonw** [comment]
I think that Rust**o**nomicon sounds slightly better.

**Comment by steveklabnik** [comment]
:+1: to o

**Comment by killercup** [comment]
While _The Advanced Rust Programming Language_ sounds like a serious manual full of technical details and concepts, _The Rustonomicon_ will prepare the reader for @Gankro's cheerful writing! :smile:

**Comment by nham** [comment]
With this change, I suppose the book will have to be translated into medieval Arabic. Many spells will need to be added as well, but this can all be work for future PRs.

**Comment by sweber6106** [comment]
"The Advanced Rust Programming Language" makes it sound like there was a prior, not-so-advanced language.  "The Rust Programming Language" is fine and tight.  Add a sub-title if absolutely required.

**Comment by jroesch** [comment]
:thumbs:

File: mk/docs.mk [hunk 1]
```diff
@@ -287,12 +287,12 @@ doc/book/index.html: $(RUSTBOOK_EXE) $(wildcard $(S)/src/doc/trpl/*.md) | doc/
...
-tarpl: doc/adv-book/index.html
+nomicon: doc/nomicon/index.html
 
-doc/adv-book/index.html: $(RUSTBOOK_EXE) $(wildcard $(S)/src/doc/tarpl/*.md) | doc/
+doc/nomicon/index.html: $(RUSTBOOK_EXE) $(wildcard $(S)/src/doc/nomicon/*.md) | doc/
 	@$(call E, rustbook: $@)
-	$(Q)rm -rf doc/adv-book
-	$(Q)$(RUSTBOOK) build $(S)src/doc/tarpl doc/adv-book
+	$(Q)rm -rf doc/nomicon
+	$(Q)$(RUSTBOOK) build $(S)src/doc/nomicon doc/nomicon
```
Comment (brson) [inline mk/docs.mk]: This translates to the ultimate URL of the doc and I wonder if 'doc/rustonomicon' would be clearer, more guessable.

File: src/doc/nomicon/README.md [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+# NOTE: This is a draft document, and may contain serious errors
+
+> Instead of the programs I had hoped for, there came only a shuddering blackness
+and ineffable loneliness; and I saw at last a fearful truth which no one had
+ever dared to breathe before — the unwhisperable secret of secrets — The fact
+that this language of stone and stridor is not a sentient perpetuation of Rust
+as London is of Old London and Paris of Old Paris, but that it is in fact
+quite unsafe, its sprawling body imperfectly embalmed and infested with queer
+animate things which have nothing to do with it as it was in compilation.
```
Comment (brson) [inline src/doc/nomicon/README.md]: This is hilarious.

File: mk/docs.mk (hunk 1 above)
Comment (Gankro) [inline mk/docs.mk]: Yes -- I personally felt `rustonomicon` was a bit too long, and also expect people to mess up the vowel after `rust`. I also don't expect much cold-typing of the URL. just autocompleteing/linking/bookmark/search.

**Comment by brson** [comment]
Well, I like this a lot. It has a lot of branding possibilities in physical publication, and thematically it has a Rust-rich backstory with the Lovecraft connection. It *does nothing at all* to stem the criticism about the casual tone in our docs, but I'm ok doubling down on the whimsy for these docs (and the casual reputation of our docs is probably more good than bad). The reference and other books can be expanded in a more straightforward technical style.

(As a recall Gankro suggesting once we might consider styling it differently to suggest the dark knowledge within, but that can be done later).

**Comment by Gankro** [comment]
(rebased)

**Comment by Gankro** [comment]
Per offline discussion I've added a subtitle for SEO/clarity/auto-complete purposes. I will be submitting a patch to our nginx config to make adv-book/rustonomicon/rustinomicon redirct to nomicon

@bors r=brson

**Comment by edunham** [comment]
Summarizing IRC discussion this morning in `#rust-internals` about the name: 

* Concerns about the name being difficult to spell and remember for non-Lovecraft fans will be addressed by adding the subtitle "The Dark Arts of Advanced and Unsafe Rust Programming"
* Because Rustdoc doesn't exactly support subtitles, the title in the pages will be "Rustonomicon: The Dark Arts of Advanced and Unsafe Rust Programming". 
* Since the name is hard to spell, the docs will live at the `/nomicon` URL and `/rustonomicon`, `/rustinomicon`, and `/adv-book` will all redirect to `/nomicon`. I'll add these redirects once the PR lands.

**Comment by Gankro** [comment]
@edunham I actually settled on an `<h4>` for the subtitle, but otherwise :+1:

**Comment by pnkfelix** [comment]
`\(;,;)/`

**Comment by GBGamer** [comment]
You've met with a terrible fate, haven't you...

**Pull request merged**